drivers/audio/audio_dma.c

```diff
@@ -0,0 +1,607 @@
+}
+
+static int audio_dma_freebuffer(struct audio_lowerhalf_s *dev,
+                                struct audio_buf_desc_s *bufdesc)
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+  struct ap_buffer_s *apb;
+
+  apb = bufdesc->u.buffer;
+  audio_dma->alloc_index--;
+  nxsem_destroy(&apb->sem);
+  kumm_free(apb);
+
+  if (audio_dma->alloc_index == 0)
+    {
+      kumm_free(audio_dma->alloc_addr);
+      audio_dma->alloc_addr = NULL;
+    }
+
+  return sizeof(struct audio_buf_desc_s);
+}
+
+static int audio_dma_enqueuebuffer(struct audio_lowerhalf_s *dev,
+                                   struct ap_buffer_s *apb)
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+  irqstate_t flags;
+
+  if (audio_dma->playback)
+    up_clean_dcache((uintptr_t)apb->samp,
+                    (uintptr_t)apb->samp + apb->nbytes);
+
+  apb->flags |= AUDIO_APB_OUTPUT_ENQUEUED;
+
+  flags = enter_critical_section();
+  dq_addlast(&apb->dq_entry, &audio_dma->pendq);
+  leave_critical_section(flags);
+
+  if (audio_dma->xrun)
+    {
+      audio_dma->xrun = false;
+      return audio_dma_resume(dev);
+    }
+
+  return OK;
+}
+
+static int audio_dma_ioctl(struct audio_lowerhalf_s *dev, int cmd,
+                           unsigned long arg)
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)dev;
+  struct ap_buffer_info_s *bufinfo;
+
+  switch (cmd)
+    {
+      /* Report our preferred buffer size and quantity */
+
+      case AUDIOIOC_GETBUFFERINFO:
+        audinfo("AUDIOIOC_GETBUFFERINFO:\n");
+        bufinfo              = (struct ap_buffer_info_s *)arg;
+        bufinfo->buffer_size = audio_dma->buffer_size;
+        bufinfo->nbuffers    = audio_dma->buffer_num;
+
+        return OK;
+
+      case AUDIOIOC_SETBUFFERINFO:
+        audinfo("AUDIOIOC_GETBUFFERINFO:\n");
+        bufinfo                = (struct ap_buffer_info_s *)arg;
+        audio_dma->buffer_size = bufinfo->buffer_size;
+        audio_dma->buffer_num  = bufinfo->nbuffers;
+        kumm_free(audio_dma->alloc_addr);
+        audio_dma->alloc_addr = NULL;
+        audio_dma->alloc_index = 0;
+
+        return OK;
+    }
+
+  return -ENOTTY;
+}
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int audio_dma_reserve(struct audio_lowerhalf_s *dev, void **session)
+#else
+static int audio_dma_reserve(struct audio_lowerhalf_s *dev)
+#endif
+{
+  return OK;
+}
+
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+static int audio_dma_release(struct audio_lowerhalf_s *dev, void *session)
+#else
+static int audio_dma_release(struct audio_lowerhalf_s *dev)
+#endif
+{
+  return OK;
+}
+
+static void audio_dma_callback(struct dma_chan_s *chan,
+                               void *arg, ssize_t len)
+{
+  struct audio_dma_s *audio_dma = (struct audio_dma_s *)arg;
+  struct ap_buffer_s *apb;
+  bool final = false;
+
+  apb = (struct ap_buffer_s *)dq_remfirst(&audio_dma->pendq);
+  if (!apb)
+    {
+      /* xrun */
+
+      DMA_PAUSE(audio_dma->chan);
```

Comment:
Where is this macro defined?

Reply:
It's defined in https://github.com/apache/nuttx/pull/8773.